### PR TITLE
MiniBatchKMeans (WIP)

### PR DESCRIPTION
**Pull request by egecimsir**
Unfortunately, not yet finished.

### CODE REVIEW

**Comment by collinleiber** [comment]
Minibatch Kmeans is missing

File: practical/DeepClustering/kmeans_Cimsir.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import torch.optim
+
+import matplotlib.pyplot as plt
+import matplotlib.cm as cm
+
+from sklearn.datasets import make_blobs
+from scipy.spatial import distance
+from torch.nn import Parameter
+
+
```
Comment (julianSchilcher) [inline practical/DeepClustering/kmeans_Cimsir.py:17]: I think it's a very good idea to include a optional visualization of the optimization process during iteration. But maybe it is enough to just visualize every 20 iterations. And maybe it would be nice to include you last fit method into the Minibatch-class to meet the sklearn convention. However, your code is good structured and nice to read!

File: practical/DeepClustering/kmeans_Cimsir.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class KMeans:
+    """
+    Standard k-Means Algorithm
+    """
+
+    def __init__(self, n_cluster, distance=distance.euclidean):
```
Comment (luftmeer) [inline practical/DeepClustering/kmeans_Cimsir.py:23]: Don't add the distance function of choice via the __init__ class method. This is bad practice, as this could result in future errors when another user wants to use another and/or custom distance function. Instead, offer a choice, such as 'euclidean' and offer a general purpose function such as distance.cdist.
Another advantage: cdist will calculate all distances between to d-Dimensional arrays, which should save compute time with larger datasets.

File: practical/DeepClustering/kmeans_Cimsir.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    def __init__(self, n_cluster):
+        super(MiniBatchKmeans, self).__init__()
+        self.kmean = KMeans(n_cluster)
+
+    def forward(self, X):
+        labels, centroids = self.kmean.fit(X, 1)
+        return centroids
+
+
+def fit(X, n_cluster, max_epoch, batch_size):
```
Comment (luftmeer) [inline practical/DeepClustering/kmeans_Cimsir.py:97]: Indentation is probably missing?
